### ISSUE TITLE
Unagi chan vanishing issue when over 250 unique websockets are sent

### ISSUE THREAD

**Issue by codygman** [jberryman/unagi-chan#17]
https://github.com/bitemyapp/websocket-shootout/pull/3#issuecomment-244559303

Basically if "step-size" from this benchmark is over 250 unagi chan starts having errors like:

```
$ ./bin/hs-websocket-server 
hs-websocket-server: writev: resource vanished (Connection reset by peer)
hs-websocket-server: ConnectionClosed
hs-websocket-server: writev: resource vanished (Broken pipe)
hs-websocket-server: writev: resource vanished (Broken pipe)
hs-websocket-server: writev: resource vanished (Broken pipe)
hs-websocket-server: writev: resource vanished (Broken pipe)
hs-websocket-server: writev: resource vanished (Broken pipe)
hs-websocket-server: writev: resource vanished (Broken pipe)
hs-websocket-server: hs-websocket-server: writev: resource vanished (Broken pipe)
writev: resource vanished (Broken pipe)
hs-websocket-server: writev: resource vanished (Broken pipe)
hs-websocket-server: writev: resource vanished (Broken pipe)
hs-websocket-server: writev: resource vanished (Broken pipe)
hs-websocket-server: writev: resource vanished (Broken pipe)
hs-websocket-server: writev: resource vanished (Broken pipe)
hs-websocket-server: writev: resource vanished (Broken pipe)
hs-websocket-server: writev: resource vanished (Broken pipe)
hs-websocket-server: writev: resource vanished (Broken pipe)
hs-websocket-server: writev: resource vanished (Broken pipe)
hs-websocket-server: writev: resource vanished (Broken pipe)
hs-websocket-server: writev: resource vanished (Broken pipe)
hs-websocket-server: writev: resource vanished (Connection reset by peer)hs-websocket-server: writev: resource vanished (Broken pipe)

hs-websocket-server: writev: resource vanished (Broken pipe)
hs-websocket-server: ConnectionClosed
```

Though, this could have been [from not increasing number of open files on the OS running the application](https://github.com/bitemyapp/websocket-shootout/pull/3#issuecomment-244563924).

Either way, I saw the performance benchmarks  on the cabal page for unagi-chan, thought you seemed very concerned with performance, and as a result would be very interested in the issue happening here.

As a bonus this hashrocket websocket example defintely counts as a realistic use case for benchmarking as you were asking for ideas on in #8.

Hopefully we can figure this out and Haskelll/unagi-chan will have a good showing in the [hashrocket websocket benchmark](https://github.com/hashrocket/websocket-shootout).

**Comment by bitemyapp** [comment]
I don't think there's an unagi-chan bug here based on what I've seen so far, so please don't go chasing a phantom.

All my profiling shows this being issues with serialization and syscalls in the websockets implementation so far.

Which makes more sense because I've used unagi-chan in production before and it has only worked a treat, whereas I've not used the Haskell websockets library in prod before.

**Comment by jberryman** [comment]
Thanks for opening the issue, and indeed maybe this is a nice realistic use case to keep an eye on! I'm going to close for now per @bitemyapp 's recommendation but please re-open if it seems to be a real issue with the library.

**Comment by bitemyapp** [comment]
@jberryman sorry to bother you, but could you take briefly a look at how we're using unagi-chan to broadcast data in the websocket benchmark to see if any obvious errors have been made?

https://github.com/sgraf812/websocket-shootout/blob/c7c0da0de72626e4a3fce94043c773ba7be090db/haskell/src/Main.hs this is the original version PR'd. If you could comment on that extraneous reader, that would help as well.

**Comment by jberryman** [comment]
At a high level this looks okay, though I'm not familiar with the websockets libraries. I didn't follow the thread so I'm not sure what you're looking for exactly (performance, or weird bugs like in ops post?). If I was debugging this, I would first remove all the websockets stuff and replace them with mocked clients communicating over Chans themselves.

Random things that came to mind: 
- it's possible you have non-allocating loops: https://ghc.haskell.org/trac/ghc/ticket/367 (e.g. a reader that always lags a writer and never blocks; maybe that "extraneous reader"). 
- What happens if connections go away? Maybe use `async` to surface exceptions to `main`. Maybe there's a thread leak hidden in those forks and forevers
- The `C.forkIO $ forever $ readChan outp` and comment don't seem correct; `InChan` and all its duplicated `OutChan` share a single array stream. `InChan` and `OutChan` just hold pointers into this array stream; when all those pointers advance beyond an array segment, that segment may be GC'd. Let me know if you think I'm misunderstanding.

**Comment by bitemyapp** [comment]
This is very helpful, thank you. It's mostly clear at this point that the slow parts are the socket connection APIs themselves. I did setup a broadcast-focused unagi-chan benchmark last night and the numbers were essentially what I expected: so fast it doesn't really show up in what the server is doing. It's possible I misunderstood, but if you nothing screams badness to you, I think the U-C side of this is fine.

>The C.forkIO $ forever $ readChan outp and comment don't seem correct;

I didn't think it was either, which is why I removed it after merging the PR as-it-was.

Thank you again, this helps build confidence in what we should focus on (network / websockets), cheers :)
